uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/main/java/org/uberfire/ext/metadata/io/BatchIndex.java

```diff
@@ -0,0 +1,259 @@
+        this.observer = checkNotNull("observer",
+                                     observer);
+        this.views = views;
+
+        this.executorService = executorService;
+    }
+
+    public void runAsync(final FileSystem fs) {
+        if (fs != null && fs.getRootDirectories().iterator().hasNext()) {
+            executorService.execute(new DescriptiveRunnable() {
+                @Override
+                public String getDescription() {
+                    return "FS BatchIndex [" + ((FileSystemId) fs).id() + "]";
+                }
+
+                @Override
+                public void run() {
+                    final AtomicBoolean indexFinished = new AtomicBoolean(false);
+                    indexEngine.beforeDispose(new Runnable() {
+                        @Override
+                        public void run() {
+                            indexDisposed.set(true);
+
+                            if (!indexFinished.get()) {
+                                indexEngine.delete(KObjectUtil.toKCluster(fs));
+                            }
+                        }
+                    });
+
+                    KCluster cluster = null;
+                    try {
+                        cluster = KObjectUtil.toKCluster(fs);
+                        indexEngine.startBatch(cluster);
+                        for (final Path root : fs.getRootDirectories()) {
+                            BatchIndex.this.run(root);
+                        }
+                        indexFinished.set(true);
+                    } catch (Exception ex) {
+                        if (!indexDisposed.get()) {
+                            logError("FileSystem Index fails. [@" + fs.toString() + "]",
+                                     ex);
+                        }
+                    }
+                    finally {
+                        if (cluster != null) {
+                            if (indexFinished.get()) {
+                                indexEngine.commit(cluster);
+                            } else {
+                                indexEngine.abort(cluster);
+                            }
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    public void runAsync(final Path root) {
+        this.executorService.execute(new DescriptiveRunnable() {
+            @Override
+            public String getDescription() {
+                return "Path BatchIndex [" + root.toString() + "]";
+            }
+
+            @Override
+            public void run() {
+                BatchIndex.this.run(root);
+            }
+        });
+    }
+
+    public void run(final Path root) {
+        run(root,
+            null);
+    }
+
+    public void run(final Path root,
+                    final Runnable callback) {
+        try {
+            if (root == null) {
+                return;
+            }
+
+            logInformation("Starting indexing of " + root.toUri() + " ...");
```

Comment:
@mbarkley I confirm that indexing now seems to work correctly both when cloning via REST and in UI as well. There's however one issue that this is showing bogus URLs in server.log when cloning via REST:

```
[org.uberfire.ext.metadata.io.BatchIndex] (Thread-155) Starting indexing of default://master@myspace/null/ ...
[org.uberfire.ext.metadata.io.BatchIndex] (Thread-155) Completed indexing of default://master@myspace/null/
[org.uberfire.ext.metadata.io.BatchIndex] (Thread-161) Starting indexing of default://master@myspace/null-1/ ...
[org.uberfire.ext.metadata.io.BatchIndex] (Thread-155) Starting indexing of default://master@myspace/null/ ...
```

This looks like minor issue which shouln't hold merging of this important fix. Should I file that separately of can that be fixed quickly?

Reply:
@jhrcek Can you please open a new issue for this?

Reply:
@mbarkley My bad, I just noticed I didn't provide "name" argument in the CloneRepositoryRequest. When I fixed that, the repo is named correctly and everything is working fine. So I'm not going to report anything related to that. Sorry for confusion.